Sources/StreamChat/Repositories/OfflineRequestsRepository.swift

```diff
@@ -0,0 +1,134 @@
+//
+// Copyright © 2022 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+typealias QueueOfflineRequestBlock = (DataEndpoint) -> Void
+typealias DataEndpoint = Endpoint<Data>
+
+extension Endpoint {
+    var withDataResponse: DataEndpoint {
+        DataEndpoint(
+            path: path,
+            method: method,
+            queryItems: queryItems,
+            requiresConnectionId: requiresConnectionId,
+            requiresToken: requiresToken,
+            body: body
+        )
+    }
+}
+
+/// OfflineRequestsRepository handles both the enqueuing and the execution of offline requests when needed.
+/// When running the queued requests, it basically passes the requests on to the APIClient, and waits for its result.
+class OfflineRequestsRepository {
+    let database: DatabaseContainer
+    let apiClient: APIClient
+
+    /// Serial queue used to enqueue pending requests one after another
+    private let retryQueue = DispatchQueue(label: "com.stream.queue-requests")
+
+    init(database: DatabaseContainer, apiClient: APIClient) {
+        self.database = database
+        self.apiClient = apiClient
+    }
+
+    /// - If the requests succeeds -> The request is removed from the pending ones
+    /// - If the request fails with a connection error -> The request is kept to be executed once the connection is back (we are not putting it back at the queue to make sure we respect the order)
+    /// - If the request fails with any other error -> We are dismissing the request, and removing it from the queue
+    func runQueuedRequests(completion: @escaping () -> Void) {
+        var pendingActions: [(String, Data)] = []
+
+        let readContext = database.backgroundReadOnlyContext
+        readContext.performAndWait {
+            pendingActions = QueuedRequestDTO.loadAllPendingRequests(context: readContext).map {
+                ($0.id, $0.endpoint)
+            }
```

Comment:
Shouldn't we read these pending requests from the retry queue as well? What happens if meanwhile, we add an offline request to the database while we are reading from it? Won't there be any concurrency issues?

Reply:
You cannot actually do that as the contexts have their own queue. Even if I move it to the retry queue, they will be then moved to the context's queue, leaving the end result as it is right now